compositor_pipeline/src/pipeline/whip_whep.rs

```diff
@@ -0,0 +1,194 @@
+    Ok((StatusCode::OK, axum::Json(json!({}))))
+}
+
+#[derive(Debug, Clone)]
+pub struct WhipInputConnectionOptions {
+    pub video_sender: Option<mpsc::Sender<PipelineEvent<EncodedChunk>>>,
+    pub audio_sender: Option<mpsc::Sender<PipelineEvent<EncodedChunk>>>,
+    pub bearer_token: Option<String>,
+    pub peer_connection: Option<Arc<RTCPeerConnection>>,
+    pub start_time_vid: Option<Instant>,
+    pub start_time_aud: Option<Instant>,
+    pub depayloader: Arc<Mutex<Depayloader>>,
+}
+
+impl WhipInputConnectionOptions {
+    pub fn get_or_initialize_elapsed_start_time(
+        &mut self,
+        track_kind: RTPCodecType,
+    ) -> Option<Duration> {
+        match track_kind {
+            RTPCodecType::Video => {
+                let start_time = self.start_time_vid.get_or_insert_with(Instant::now);
+                Some(start_time.elapsed())
+            }
+            RTPCodecType::Audio => {
+                let start_time = self.start_time_aud.get_or_insert_with(Instant::now);
+                Some(start_time.elapsed())
+            }
+            _ => None,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct WhipWhepState {
+    pub input_connections: Arc<Mutex<HashMap<InputId, WhipInputConnectionOptions>>>,
+}
+
+impl WhipWhepState {
+    pub fn new() -> Arc<Self> {
+        Arc::new(WhipWhepState {
+            input_connections: Arc::from(Mutex::new(HashMap::new())),
+        })
+    }
+
+    pub fn get_input_connection_options(
+        &self,
+        input_id: InputId,
+    ) -> Result<WhipInputConnectionOptions, WhipServerError> {
+        let connections = self.input_connections.lock().unwrap();
+        if let Some(connection) = connections.get(&input_id) {
+            if let Some(peer_connection) = connection.peer_connection.clone() {
+                warn!("There is another stream streaming for given input {input_id:?}");
+                if peer_connection.connection_state() == RTCPeerConnectionState::Connected {
+                    return Err(WhipServerError::InternalError(format!(
+                        "There is another stream streaming for given input {input_id:?}"
+                    )));
+                }
+            }
+            Ok(connection.clone())
+        } else {
+            Err(WhipServerError::NotFound(format!(
+                "InputID {input_id:?} not found"
+            )))
+        }
+    }
+
+    pub async fn update_peer_connection(
+        &self,
+        input_id: InputId,
+        peer_connection: Arc<RTCPeerConnection>,
+    ) -> Result<(), WhipServerError> {
+        let mut connections = self.input_connections.lock().unwrap();
+        if let Some(connection) = connections.get_mut(&input_id) {
+            connection.peer_connection = Some(peer_connection);
+            Ok(())
+        } else {
+            Err(WhipServerError::InternalError(
+                "Peer connection initialization error".to_string(),
```

Comment:
This error description is misleading, this can happen if there if input with that id does not exist.